Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs

```diff
@@ -1,32 +1,152 @@
 {
+    /// <summary>
+    /// Contains all the server-side logic for reagent dispensers.
+    /// <seealso cref="ReagentDispenserComponent"/>
+    /// </summary>
     [UsedImplicitly]
-    public sealed class ReagentDispenserSystem : SharedReagentDispenserSystem
+    public sealed class ReagentDispenserSystem : EntitySystem
     {
+        [Dependency] private readonly AudioSystem _audioSystem = default!;
+        [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+        [Dependency] private readonly ItemSlotsSystem _itemSlotsSystem = default!;
+        [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+        [Dependency] private readonly EntityManager _entities = default!;
         public override void Initialize()
         {
             base.Initialize();
-            SubscribeLocalEvent<ReagentDispenserComponent, ComponentStartup>((_, comp, _) => comp.UpdateUserInterface());
-            SubscribeLocalEvent<ReagentDispenserComponent, SolutionChangedEvent>((_, comp, _) => comp.UpdateUserInterface());
-            SubscribeLocalEvent<ReagentDispenserComponent, EntInsertedIntoContainerMessage>((_, comp, _) => comp.UpdateUserInterface());
-            SubscribeLocalEvent<ReagentDispenserComponent, EntRemovedFromContainerMessage>((_, comp, _) => comp.UpdateUserInterface());
+
+            SubscribeLocalEvent<ReagentDispenserComponent, ComponentStartup>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, SolutionChangedEvent>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, EntInsertedIntoContainerMessage>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, EntRemovedFromContainerMessage>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, BoundUIOpenedEvent>((_, comp, _) => UpdateUiState(comp));
             SubscribeLocalEvent<ReagentDispenserComponent, GotEmaggedEvent>(OnEmagged);
+
+            SubscribeLocalEvent<ReagentDispenserComponent, ReagentDispenserSetDispenseAmountMessage>(OnSetDispenseAmountMessage);
+            SubscribeLocalEvent<ReagentDispenserComponent, ReagentDispenserDispenseReagentMessage>(OnDispenseReagentMessage);
+            SubscribeLocalEvent<ReagentDispenserComponent, ReagentDispenserClearContainerSolutionMessage>(OnClearContainerSolutionMessage);
+        }
+
+        private void UpdateUiState(ReagentDispenserComponent reagentDispenser)
+        {
+            if (!HasComp<ItemSlotsComponent>(reagentDispenser.Owner))
+                return;
+
+            var outputContainer = _itemSlotsSystem.GetItem(reagentDispenser.Owner, SharedReagentDispenser.OutputSlotName);
```

Comment:
Do I always need to check if an entity has `ItemSlotsComponent` before calling `_itemSlotsSystem.GetItem()`?
Or should `GetItem()` maybe be changed to not throw an exception and just return null it the component is missing (`Resolve(... logMissing: false)`)?

Reply:
changing it to `logMissing: false` seems reasonable here, though in that case the API should probably be `TryGetItem` instead

Reply:
Changed it to `logMissing: false`. All the `TryGetThing()` functions seem to return `bool` and have the thing it gets as an out parameter. So I left the name as is for that reason.

Reply:
I decided on `GetItemOrNull()` since that seems more common for things that return nullable.